### ISSUE TITLE
Obfuscated Trait Detection, Thresholds and Recursion

### ISSUE THREAD

**Issue by c3rb3ru5d3d53c** [c3rb3ru5d3d53c/binlex#7]
- :heavy_check_mark:  Cyclomatic Complexity
- :heavy_check_mark: Basic Block Size in Bytes
- :heavy_check_mark: Basic Block Instruction Count
- :heavy_check_mark: Function Size
- :heavy_check_mark: Average Instructions per Block
- :heavy_check_mark: Use `cs_disasm_iter()` for improved speed and control over program counter
- :heavy_check_mark: Pretty Print
- :heavy_check_mark:  Recursive Decompilation by Instruction (fine-tuned control over exceptions)
- :heavy_check_mark: Wildcard Scalars

Allows the user to fine-tune their output, calculating these is best left to decompiler stage especially cyclomatic complexity.

```text
[
  {
    "average_instructions_per_block": 3,
    "blocks": 1,
    "bytes": "01 c3 29 c6 75 c1",
    "bytes_entropy": 0,
    "bytes_sha256": "5776a6a5e142981e2848b93a068268018809b786e310fca8b142cadd724f6f9a",
    "instructions": 3,
    "offset": 337,
    "size": 6,
    "type": "block"
  },
  {
    "average_instructions_per_block": 10,
    "blocks": 15,
    "bytes": "fc e8 8f 00 00 00 60 89 e5 31 d2 64 8b 52 30 8b 52 0c 8b 52 14 31 ff 8b 72 28 0f b7 4a 26 31 c0 ac 3c 61 7c 02 2c 20 c1 cf 0d 01 c7 49 75 ef 52 57 8b 52 10 8b 42 3c 01 d0 8b 40 78 85 c0 74 4c 01 d0 50 8b 58 20 8b 48 18 01 d3 85 c9 74 3c 49 8b 34 8b 01 d6 31 ff 31 c0 c1 cf 0d ac 01 c7 38 e0 75 f4 03 7d f8 3b 7d 24 75 e0 58 8b 58 24 01 d3 66 8b 0c 4b 8b 58 1c 01 d3 8b 04 8b 01 d0 89 44 24 24 5b 5b 61 59 5a 51 ff e0 58 5f 5a 8b 12 e9 80 ff ff ff 5d 68 33 32 00 00 68 77 73 32 5f 54 68 4c 77 26 07 89 e8 ff d0 b8 90 01 00 00 29 c4 54 50 68 29 80 6b 00 ff d5 6a 0a 68 5d b8 d8 22 68 02 00 11 5c 89 e6 50 50 50 50 40 50 40 50 68 ea 0f df e0 ff d5 97 6a 10 56 57 68 99 a5 74 61 ff d5 85 c0 74 0a ff 4e 08 75 ec e8 67 00 00 00 6a 00 6a 04 56 57 68 02 d9 c8 5f ff d5 83 f8 00 7e 36 8b 36 6a 40 68 00 10 00 00 56 6a 00 68 58 a4 53 e5 ff d5 93 53 6a 00 56 53 57 68 02 d9 c8 5f ff d5 83 f8 00 7d 28 58 68 00 40 00 00 6a 00 50 68 0b 2f 0f 30 ff d5 57 68 75 6e 4d 61 ff d5 5e 5e ff 0c 24 0f 85 70 ff ff ff e9 9b ff ff ff 01 c3 29 c6 75 c1 c3",
    "bytes_entropy": 0,
    "bytes_sha256": "ab8e5368e7965b1520f44ab6b7b66ebdf9c9d203b730e444eed758856a07cdb3",
    "instructions": 150,
    "offset": 0,
    "size": 344,
    "type": "function"
  }
]
```

Thresholds with `jq` make hunting easy with a query language:
```
build/binlex -m raw:x86 -i tests/raw/raw.x86 | jq -r '.[] | select(.type == "block" and .size < 32 and .size > 0) | .bytes'
2c 20 c1 cf 0d 01 c7 49 75 ef
52 57 8b 52 10 8b 42 3c 01 d0 8b 40 78 85 c0 74 4c
01 d0 50 8b 58 20 8b 48 18 01 d3 85 c9 74 3c
49 8b 34 8b 01 d6 31 ff 31 c0 c1 cf 0d ac 01 c7 38 e0 75 f4
03 7d f8 3b 7d 24 75 e0
58 5f 5a 8b 12 e9 80 ff ff ff
ff 4e 08 75 ec
e8 67 00 00 00 6a 00 6a 04 56 57 68 02 d9 c8 5f ff d5 83 f8 00 7e 36
e9 9b ff ff ff
01 c3 29 c6 75 c1
```

:heavy_check_mark:  To achieve easier management of strings move to `std::string` and `json` instead of `char *`.

Trait format will change thus should be a `minor` version bump.

Article with research:
[obf.pdf](https://github.com/c3rb3ru5d3d53c/binlex/files/7581072/obf.pdf)

References:
- https://synthesis.to/2021/08/10/obfuscation_detection.html
- https://github.com/alexander-hanel/capstool/blob/master/capstool/capstool.py (recursion)

**Comment by c3rb3ru5d3d53c** [comment]
Making some strides here, output is close to Cutter's output.

![image](https://user-images.githubusercontent.com/79795199/143440441-56de6390-e352-474f-9156-fbc818e57485.png)

Still a linear decompiler but we do have decent control over the program counter with `cs_disasm_iter()` now. To make a note that was not in any capstone docs I was reading, if `cs_disasm_iter()` encounters an invalid instruction you need to use `memmove()` to move the memory it points to yourself.

**Comment by c3rb3ru5d3d53c** [comment]
```bash
build/binlex -m raw:x86 -i tests/raw/raw.emotet.x86 --pretty | jq -c '[.[] | select(.type == "function" and .cyclomatic_complexity > 20) | .trait]'
```

**Comment by c3rb3ru5d3d53c** [comment]
Next step is to get the same output for `raw:cil` mode.

**Comment by c3rb3ru5d3d53c** [comment]
Closed, moving recursive decompiler to `v1.1.1` patch release.
